src/Form/BasicSetup.php

```diff
@@ -0,0 +1,524 @@
+      //    $codes = $recovery->getCodes();
+      //
+      //    $output = ['#theme' => 'item_list', '#items' => $codes];
+      //    $output = \Drupal::service('renderer')->render($output);
+      //    $output .= Link::fromTextAndUrl(t('Return to account TFA overview'), Url::fromRoute('tfa.overview', ['user' => $account->id()]))->toString();
+      //    $form['output'] = array(
+      //      '#type' => 'markup',
+      //      '#markup' => $output,
+      //    );
+      //    // Return early.
+      //    return $form;
+      //
+      //  default:
+      //    break;
+      // }
+      // Provide skip button under full setup.
+      if (isset($storage['full_setup']) && count($storage['steps']) > 1) {
+        $count = count($storage['steps_left']);
+        $form['actions']['skip'] = array(
+          '#type' => 'submit',
+          '#value' => $count > 0 ? t('Skip') : t('Skip and finish'),
+          '#limit_validation_errors' => array(),
+          '#submit' => array('::cancelForm'),
+        );
+      }
+      // Provide cancel button on first step or single steps.
+      else {
+        $form['actions']['cancel'] = array(
+          '#type' => 'submit',
+          '#value' => t('Cancel'),
+          '#limit_validation_errors' => array(),
+          '#submit' => array('::cancelForm'),
+        );
+      }
+      // Record the method in progress regardless of whether in full setup.
+      $storage['step_method'] = $method;
+    }
+    $form_state->setStorage($storage);
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    $user = User::load(\Drupal::currentUser()->id());
+    $storage = $form_state->getStorage();
+    $values = $form_state->getValues();
+    $account = $form['account']['#value'];
+    if (isset($values['current_pass'])) {
+      // Allow administrators to change TFA settings for another account.
+      if ($account->id() != $user->id() && $user->hasPermission('administer users')) {
+        $account = $user;
+      }
+      $current_pass = \Drupal::service('password')->check(trim($form_state->getValue('current_pass')), $account->getPassword());
+      if (!$current_pass) {
+        $form_state->setErrorByName('current_pass', t("Incorrect password."));
+      }
+
+      // Check password. (from user.module user_validate_current_pass()).
+      //      require_once DRUPAL_ROOT . '/' . variable_get('password_inc', 'includes/password.inc');
+      //      $current_pass = user_check_password($values['current_pass'], $account);
+      //      if (!$current_pass) {
+      //        form_set_error('current_pass', t("Incorrect password."));
+      //      }.
+      return;
+    }
+    // Handle first step of SMS setup.
+    elseif (isset($values['sms_number'])) {
+      // Validate number.
+      $number = $values['sms_number'];
+      $number_errors = tfa_basic_valid_number($number);
```

Comment:
This should be moved to a service or function of the class and not a procedural callback